### ISSUE TITLE
`Walker` trait? (akin to Iterator, but for Walker types)

### ISSUE THREAD

**Issue by mitchmindtree** [petgraph/petgraph#13]
I quite often come up with the desire to "filter"/"count"/etc the node edge pairs produced by the Walker types - perhaps it would be beneficial to abstract the `Walker` behaviour into a trait that behaves similarly to `Iterator` (with the difference being that `next` takes a graph by ref)?

Something like (without all the generics):

``` rust
pub trait Walker {
    fn next(&mut self, graph: &Graph) -> Option<(EdgeIndex, NodeIndex)>;

    fn next_node(&mut self, graph: &Graph) -> Option<NodeIndex> {
        ...
    }
    fn next_edge(&mut self, graph: &Graph) -> Option<EdgeIndex> {
        ...
    }

    // Provide heaps of other useful adapators like Filter, Map, etc
    // ...
}
```

I might have a go at drafting something up in daggy.

**Comment by mitchmindtree** [comment]
Ok, I've had a go at implementing a [**Walker** trait](https://github.com/mitchmindtree/daggy/blob/walker/src/walker.rs) and have [written some tests](https://github.com/mitchmindtree/daggy/blob/walker/tests/walkers.rs) which demonstrate their behaviour.

I've found that, for graph traversal, walkers are strictly more flexible than iterators - not only does a walker not borrow the graph, it is also easily converted to an iterator while an iterator may not (as easily) be converted to a walker.

In summary, I think the trait could be useful for making Graph's neighbour-related method names a little more consistent, filling some missing holes in functionality while also allowing to remove some methods that would become unnecessary. (Beware: the following contains some ideas that would cause heavily breaking changes :bomb: ).

Let's say that we were to change the `.neighbors`, `.neighbors_directed` and `.neighbors_undirected` methods to return `Walker` types (more flexible than the current iterator yielding `NodeIndex`s) while adding an additional `.neighbors_directed_both` walker to complete our set. These four walkers should be all that is needed for visiting neighbouring nodes.

Name Distinction
------------------------

The name distinction between "neighbors" and "edges" would become clearer:
- A "neighbor" would strictly describe a neighbouring (edge, node) pair. i.e.
    - `.neighbors(idx).next(&g)` yields `(EdgeIndex, NodeIndex)`
    - `.neighbors(idx).iter(&g)` creates an iterator yielding `(EdgeIndex, NodeIndex)`
    - `.neighbors(idx).iter_weights(&g)` creates an iterator yielding `(&E, &N)`
- An "edge" would strictly describe the edge part of a "neighbor". i.e.
    - `.neighbors(idx).next_edge(&g)` yields `EdgeIndex`
    - `.neighbors(idx).iter(&g).edges()` creates an iterator yielding `EdgeIndex`
    - `.neighbors(idx).iter_weights(&g).edges()` creates an iterator yielding `&E`
- A "node" would strictly describe the node part of a "neighbor". i.e.
    - `.neighbors(idx).next_node(&g)` yields `NodeIndex`
    - `.neighbors(idx).iter(&g).nodes()` creates an iterator yielding `NodeIndex`
    - `.neighbors(idx).iter_weights(&g).nodes()` creates an iterator yielding `&N`

Minimalising API and Improving consistency
------------------------------------------------------------

- The original `.neighbors(idx)` could be described as `.neighbors(idx).iter(&g).nodes()` (same for directed and undirected variants).
- `.edges(idx)` could become `.neighbors(idx).iter(&g)`, or a more exact translation would be `.neighbors(idx).iter(&g).map(|(e, n)| (n, &g[e]))` (same for directed variant).
- `.edges_both(idx)` could become `.neighbors_directed_both(idx)` instead yielding a (more flexible) walker, which could achieve the original functionality in the same way as described in the previous point.
- `.walk_edges_directed(idx, dir)` would no longer be needed, as `.neighbors_directed(idx, dir)` would take its place.

This would cut our methods down from the original:
- `neighbors`
- `neighbors_directed`
- `neighbors_undirected`
- `edges`
- `edges_directed`
- `edges_both`
- `walk_edges_directed`

to:
- `neighbors`
- `neighbors_directed`
- `neighbors_undirected`
- `neighbors_directed_both`


Anyway, I know you're busy atm, but it'd be great to get your thoughts when you next get a chance :smiley_cat: I'd be happy to put together a PR if you decided `Walker` would be a suitable addition. With or without the breaking changes suggested above, it's probably a much better fit upstream in petgraph than it is in daggy, as it is applicable to *any* graph type (not just DAGs). Otherwise, I'll just leave it in the daggy crate for now.

**Comment by bluss** [comment]
It's really nice.

Walker vs iterator derived from it does have one example in Dfs vs DfsIter. Dfs is an example too of a walker that is not tied to one Graph type, but allows several different. It's used in the scc algorithm to use the same Dfs walker with both `&Graph` and `&Reversed(&Graph)`.

**Comment by mitchmindtree** [comment]
Ahh indeed. It should be pretty easy to move the associated `Graph` type to a generic input type `G`, so that the `Walker` trait could be implemented uniquely for different graph types too.

**Comment by bluss** [comment]
Dfs isnt' that kind of Walker anyway -- it doesn't do neighbors, just a sequence of nodes.

**Comment by mitchmindtree** [comment]
I've been noticing a few cases recently where it would be nice to be able to pass a `Dfs` or `Bfs` a specific type of `Walker` to use when walking the neighbours of the current node rather than just visiting all of them as is [currently the case](http://bluss.github.io/petulant-avenger-graphlibrary/doc/src/petgraph/visit.rs.html#361).

For example, I often want to do a depth first search of my graph where I *only* visit neighbours who are connected via a specific kind of `Edge`. In this case, it would be nice if I could have the `Dfs` produce a `Filter` `Walker` type whose predicate returns true for those `Edges` I want to visit and false for those I do not.

Perhaps this could be done by offering a way to pass the `Dfs` (or `Bfs`) a function to use to produce a `Walker` type from some given `NodeIndex` and `Graph`, where the function is some `FnMut(&Graph, NodeIndex<Ix>) -> W` and `W` is some `Walker`. The default function type could simply produce a regular walker.

**Comment by bluss** [comment]
Oh god, I just realized that maybe Walkers can get rid of all the boring HKT-lifetime-trait issues too.

Anyway, thanks for all the input. We need a more generic Dfs visitor in general, something with event hooks like this maybe: [`depth_first_search` from boost](http://www.boost.org/doc/libs/1_59_0/libs/graph/doc/depth_first_search.html).  An edge and node filter sounds very reasonable, too.

**Comment by bluss** [comment]
Can you explain how you use the recursive walker? It returns an edge index and a node index, I suppose that's the edge used to get to the next node (the node index). Do you use this for "drilling down" the DAG?

**Comment by mitchmindtree** [comment]
Yep "drilling down" is right, here's an [example](https://github.com/mitchmindtree/conrod/blob/graphics_widgets/src/graph/mod.rs#L477) from conrod and [some usage](https://github.com/mitchmindtree/conrod/blob/graphics_widgets/src/graph/algo.rs#L249).

**Comment by bluss** [comment]
There is now a Walker trait in 0.4.0, but it's nowhere near the grand vision.

**Comment by lilith** [comment]
It seems to have been made private in 0.5.0, which broke my code.

**Comment by bluss** [comment]
`petgraph::visit::Walker` is still public, but if it doesn't work, please file an issue with the details thanks

**Comment by RaoulLuque** [comment]
I will close this as resolved, since the feature is present :)
